drivers/flash/nrf_qspi_nor.c

```diff
@@ -519,25 +524,51 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 
 	qspi_wait_for_completion(dev, res);
 
-	return qspi_get_zephyr_ret_code(res);
+	int rc = qspi_get_zephyr_ret_code(res);
+
+	if ((rc == 0) && (dest != dptr)) {
+		memcpy(dptr, dest, dlen);
+	}
+
+	return rc;
 }
 
 static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
 			  size_t size)
 {
-	if (!src) {
-		return -EINVAL;
-	}
+	const void *sptr = src;
+	size_t dlen = size;
+	u8_t __aligned(4) buf[4];
 
-	/* write size must be non-zero multiple of 4 bytes */
-	if (((size % 4U) != 0) || (size == 0)) {
+	if (!src || (size == 0)) {
 		return -EINVAL;
 	}
+
 	/* address must be 4-byte aligned */
 	if ((addr % 4U) != 0) {
 		return -EINVAL;
 	}
 
+	/* Since the QSPI driver requires data to be at least 4 bytes we need
+	 * to use a 4 byte buffer for writes smaller than 4 bytes.
+	 */
+	if (size < 4U) {
+		src = buf;
+		size = sizeof(buf);
+		/* read out the whole word so that unchanged data can be
+		 * written back
+		 */
+		nrfx_err_t res = nrfx_qspi_read(buf, size, addr);
```

Comment:
This read is outside the lock.  For correctness it probably should be inside the same lock that's used for the write: a context switch could allow another thread to come in and change the value that was read into `buf` so the following write contains different data in areas that weren't supposed to change.

https://github.com/pabigot/fw-nrfconnect-zephyr/commits/pr/305 has a possible fix.

Reply:
Which isn't correct because it doesn't properly handle the potential for multiple writes in the loop below.

The locking infrastructure doesn't seem prepared to handle multiple nrfx_spi invocations for a single Zephyr-level transaction.  I'm not sure who should tackle this problem, though I can give it a shot.

Reply:
I believe the branch should work.  It passes a simple test, but `samples/subsys/fs/littlefs` doesn't work on the master branch so I can't confirm the changes work.
```
===================================================================
starting test - test_lfs_basic
clearing partition /sml
Error accessing flash area 5 [-2]

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/fs/littlefs/src/test_lfs_basic.c:44: )
failed to wipe partition
FAIL - test_lfs_basic
```
This may be my environment; I don't use ncs generally.

Reply:
Oh I think my intention was to use the qspi read

Reply:
OK, updated again at https://github.com/pabigot/fw-nrfconnect-zephyr/commits/pr/305.  There're multiple upstream fixes in zephyrproject-rtos/zephyr#25661 that are part of that branch.

Reply:
So added your fix and most of it works fine. Just found out that MCUBoot also has some buffers it passes in that are not word aligned. So I must have been lucky and got word aligned buffers before.

I'll add a temp hack to MCUBoot